File: simbus/simulation/faults.py
"""Fault injection types.

Faults are short-lived overrides applied by the SimulationEngine on each tick.
They expire automatically when their `remaining_s` reaches zero.
The fault API is exposed via REST endpoints in the simulation router.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum


class FaultType(StrEnum):
    """Types of faults that can be injected."""

    spike = "spike"  # Force a register to an extreme value for a duration
    freeze = "freeze"  # Stop updating a register (stuck sensor)
    dropout = "dropout"  # Set register to 0 (loss of signal)
    alarm = "alarm"  # Force a named coil to True, bypassing trigger evaluation
    noise_amplify = "noise_amplify"  # Multiply noise std_dev by a factor


@dataclass
class ActiveFault:
    """An active fault currently affecting the simulation."""

    fault_type: FaultType
    register_name: str | None  # None means device-wide (e.g. dropout)
    value: float | None  # spike target, noise factor, etc.
    duration_s: float
    remaining_s: float

File: crates/engine/src/faults.rs
//! Active fault records.

use spec::FaultType;

/// A TTL override applied on each tick.
#[derive(Debug, Clone, PartialEq)]
pub struct ActiveFault {
    /// Kind.
    pub fault_type: FaultType,
    /// Register or coil name; `None` means device-wide.
    pub register_name: Option<String>,
    /// Spike target or noise factor.
    pub value: Option<f64>,
    /// Original duration.
    pub duration_s: f64,
    /// Time remaining.
    pub remaining_s: f64,
}

impl ActiveFault {
    /// Build a fault with remaining time equal to duration.
    #[must_use]
    pub fn new(
        fault_type: FaultType,
        register_name: Option<String>,
        value: Option<f64>,
        duration_s: f64,
    ) -> Self {
        Self {
            fault_type,
            register_name,
            value,
            duration_s,
            remaining_s: duration_s,
        }
    }

    /// Map key used in the fault table.
    #[must_use]
    pub fn key(&self) -> String {
        self.register_name
            .clone()
            .unwrap_or_else(|| "_device".to_owned())
    }
}
